plugins/performance-lab/includes/admin/plugins.php

```diff
@@ -54,61 +54,94 @@ function perflab_query_plugin_info( string $plugin_slug ) {
 		)
 	);
 
+	$has_errors = false;
+	$plugins    = array();
+
 	if ( is_wp_error( $response ) ) {
-		return new WP_Error(
-			'api_error',
-			sprintf(
-				/* translators: %s: API error message */
-				__( 'Failed to retrieve plugins data from WordPress.org API: %s', 'performance-lab' ),
-				$response->get_error_message()
-			)
+		$plugins[ $plugin_slug ] = array(
+			'error' => array(
+				'code'    => 'api_error',
+				'message' => sprintf(
+					/* translators: %s: API error message */
+					__( 'Failed to retrieve plugins data from WordPress.org API: %s', 'performance-lab' ),
+					$response->get_error_message()
+				),
+			),
 		);
+
+		$has_errors = true;
 	}
 
 	// Check if the response contains plugins.
-	if ( ! ( is_object( $response ) && property_exists( $response, 'plugins' ) ) ) {
-		return new WP_Error( 'no_plugins', __( 'No plugins found in the API response.', 'performance-lab' ) );
+	if ( ! $has_errors && ! ( is_object( $response ) && property_exists( $response, 'plugins' ) ) ) {
+		$plugins[ $plugin_slug ] = array(
+			'error' => array(
+				'code'    => 'no_plugins',
+				'message' => __( 'No plugins found in the API response.', 'performance-lab' ),
+			),
+		);
+
+		$has_errors = true;
 	}
 
-	$plugins      = array();
-	$plugin_queue = perflab_get_standalone_plugins();
+	if ( ! $has_errors && is_object( $response ) && property_exists( $response, 'plugins' ) ) {
+		$plugin_queue = perflab_get_standalone_plugins();
 
-	// Index the plugins from the API response by their slug for efficient lookup.
-	$all_performance_plugins = array_column( $response->plugins, null, 'slug' );
+		// Index the plugins from the API response by their slug for efficient lookup.
+		$all_performance_plugins = array_column( $response->plugins, null, 'slug' );
 
-	// Start processing the plugins using a queue-based approach.
-	while ( count( $plugin_queue ) > 0 ) { // phpcs:ignore Squiz.PHP.DisallowSizeFunctionsInLoops.Found
-		$current_plugin_slug = array_shift( $plugin_queue );
+		// Start processing the plugins using a queue-based approach.
+		while ( count( $plugin_queue ) > 0 ) { // phpcs:ignore Squiz.PHP.DisallowSizeFunctionsInLoops.Found
+			$current_plugin_slug = array_shift( $plugin_queue );
 
-		if ( isset( $plugins[ $current_plugin_slug ] ) ) {
-			continue;
-		}
+			if ( isset( $plugins[ $current_plugin_slug ] ) ) {
+				continue;
+			}
 
-		if ( ! isset( $all_performance_plugins[ $current_plugin_slug ] ) ) {
-			return new WP_Error(
-				'plugin_not_found',
-				__( 'Plugin not found in WordPress.org API response.', 'performance-lab' )
-			);
+			if ( ! isset( $all_performance_plugins[ $current_plugin_slug ] ) ) {
+				// Cache the fact that the plugin was not found.
+				$plugins[ $current_plugin_slug ] = array(
+					'error' => array(
+						'code'    => 'plugin_not_found',
+						'message' => __( 'Plugin not found in API response.', 'performance-lab' ),
+					),
+				);
+				continue;
```

Comment:
Due to the recent commit [ff1b440](https://github.com/b1ink0/performance/tree/ff1b4409a9e3faa819ab3b5a2796d972643f3080) this could cause incorrect behavour if `$plugin_slug` is not found in the `all_performance_plugins` it will add error for that `$plugin_slug` as we are now using the `$has_errors` flag which will not be set to `true` in this case. The below logic will fail. 
Solution could be to check it using the `isset( $plugins[ $plugin_slug ]['error'] )` condition but it will defect the usage of the `$has_errors` flag.
```php
	if ( $has_errors ) {
		set_transient( $transient_key, $plugins, MINUTE_IN_SECONDS );

		return new WP_Error(
			$plugins[ $plugin_slug ]['error']['code'],
			$plugins[ $plugin_slug ]['error']['message']
		);
	}
```

Reason for the `$has_errors` flag not setting: As the errors could occur for other plugin ( not `$plugin_slug` ) currently code will cache the error but will return the requested `$plugin_slug` information.

Question: Should we return error even if the error is encontered for the other plugin ( not `$plugin_slug` ) ?
cc: @felixarntz

Reply:
See my reply in https://github.com/WordPress/performance/pull/1694#discussion_r1860904156 - I think this problem would become irrelevant if we checked that the plugin slug is among our performance plugins in the beginning of the function?

Reply:
Revisiting this: Why should we _not_ set the `$has_errors` flag to `true` here? I think we should, given this is an error too.

Reply:
> Reason for the $has_errors flag not setting: As the errors could occur for other plugin ( not $plugin_slug ) currently code will cache the error but will return the requested $plugin_slug information.

> Question: Should we return error even if the error is encountered for the other plugin ( not $plugin_slug ) ?

This above is reason. And question related to your question.

Reply:
I asked because I couldn't follow your initial explanation. Can you clarify what you mean?

This error is only set for the specific plugin slug that is not found. So it would not affect another plugin's return value. I think what you're referring to is that the function right now always returns a `WP_Error` when there are any errors, but that's not right, so it needs to be fixed. The `$has_errors` flag should only be used to change the transient duration.

After the transient is set, we need to check the individual `$plugins[ $plugin_slug ]` entry and only return a `WP_Error` if _that_ plugin has an error in it.

Reply:
Ohh thanks for clarification I misunderstood what was suggested before. We should set `$has_errors` if any errors is encountered and use it only for changing the transient duration and only return error if requested `$plugins[ $plugin_slug ]` has a error.